writeup/src/body.lhs

```diff
@@ -4,149 +4,160 @@
-\emph{Well-typed}, here, refers to the fact that a |Patch| is guaranteed, by
-construction, to generate well-typed elements once it is applied. This
-is not the case for the approaches~\cite{Asenov2017,Falleri2014} using
-\texttt{xml} or \texttt{json} to represent their abstract syntax
-trees, where the result of an arbitrary patch could be an ill formed tree.
-
+Software Version Control Systems are an essential tool in modern
+software development. Platforms such as Github and Bitbucket are only
+possible due to the widespread adoption of tools such as git,
+mercurial and darcs, that enable multiple developers to collaborate
+effectively.  At the heart of all these tools is the UNIX \texttt{diff}
+utility~\cite{McIlroy1976}, that computes a patch between two versions
+of a file.  The \texttt{diff} utility compares files on a line-by-line
+basis and attempts to share lines between its source and destination
+whenever possible. While this works well in practice, this fixed
+granularity of change cannot describe common refactorings---such as
+variable renaming---effectively. Representing all patches on the basis
+of the lines that have been changed may therefore lead to spurious
+\emph{merge conflicts} when combining the work of different
+developers.
+%Wouter: or maximize copies?
+
+In this paper we will present an efficient datatype generic algorithm
+to compute the difference between two algebraic datatypes, while
+guaranteeing to share subtrees whenever possible. In particular, this
+algorithm can be instantiated to the abstract syntax tree of a
+programming language---thereby enabling two changes to the same line
+of code to be merged, provided they do not modify the same parts of
+the underlying tree. We have implemented our algorithm in Haskell,
+using the datatype generic libraries it provides. %todo citation to mr-sop?
+
+Before making stating the novel contributions of this paper, we will
+start by giving a specification of the problem that we aim to solve.
+In general, we intend to compute the difference between two trees of
+type |a|, and represent these changes in some type, |Patch a|. There
+are two functions, |diff| and |apply|, that manipulate patches. The
+|diff| function computes the differences between two values of type
+|a|; wherease the |apply| function attempts to transform a value of
+type |a| according to the information stored in its argument patch:
 \begin{myhs}
 \begin{code}
 diff   :: a -> a -> Patch a
 apply  :: Patch a -> a -> Maybe a 
 \end{code}
 \end{myhs}
+Note that the |apply| function may fail, for example, when attempting
+to delete data that is not present. Yet when it succeeds, the |apply|
+function must return a value of type |a|. This may seem like an
+obvious design choice, but this property does not hold for the
+approaches~\cite{Asenov2017,Falleri2014} using \texttt{xml} or
+\texttt{json} to represent their abstract syntax trees, where the
+result of applying a patch may produce ill-formed tree.
+
```

Comment:
I like how the introduction of untyped approaches feel more natural in the dialogue